doc: net: Add HTTP client library page (#46791)
zephyrproject-rtos/zephyr: doc/connectivity/networking/api/http.rst

.. _http_interface:

> HTTP
> ####

Reviewer: ```suggestion
HTTP client
###########
```

.. contents::
    :local:
    :depth: 2

Overview
********

> Zephyr provides a HTTP client library which allows to send HTTP requests and
> parse HTTP responses. The library communicates over socket API, it does not
> however create sockets on its own. The application is responsible for creating
> a socket and passing it to the library. Therefore, depending on the application
> needs, the library can communicate either over plain TCP socket (HTTP) or
> over TLS socket (HTTPS).

Reviewer 2: ```suggestion
Zephyr provides an HTTP client library which allows to send HTTP requests and
parse HTTP responses. The library communicates over the sockets API, it does not
however create sockets on its own. The application is responsible for creating
a socket and passing it to the library. Therefore, depending on the application's
needs, the library can communicate either over a plain TCP socket (HTTP) or
over a TLS socket (HTTPS).
```

> Zephyr provides an HTTP client library which allows to send HTTP requests and
> parse HTTP responses. The library communicates over the sockets API, it does not
> however create sockets on its own. The application is responsible for creating
> a socket and passing it to the library. Therefore, depending on the application's
> needs, the library can communicate either over a plain TCP socket (HTTP) or
> over a TLS socket (HTTPS).

Reviewer: ```suggestion
The HTTP client library allows you to send HTTP requests and
parse HTTP responses. The library communicates over the sockets
API but it does not create sockets on its own.

The application must be responsible for creating a socket and passing it to the library.
Therefore, depending on the application's needs, the library can communicate over
either a plain TCP socket (HTTP) or a TLS socket (HTTPS).
```

Sample Usage
************

> The overall API of the HTTP client library is simple and consists of a single
> function. It's important to setup the request structure correctly, for example:

Reviewer: ```suggestion
The API of the HTTP client library has a single function.

The following is an example of a request structure created correctly:
```

.. code-block:: c

    struct http_request req = { 0 };
    static uint8_t recv_buf[512];

    req.method = HTTP_GET;
    req.url = "/";
    req.host = "localhost";
    req.protocol = "HTTP/1.1";
    req.response = response_cb;
    req.recv_buf = recv_buf;
    req.recv_buf_len = sizeof(recv_buf);

    /* sock is a file descriptor referencing a socket that has been connected
     * to the HTTP server.
     */
    ret = http_client_req(sock, &req, 5000, NULL);

> If the server sent a response for the request, it will be provided to the
> application through the response callback registered in the request structure.
> The response may be provided in chunks, therefore the application shall be
> prepared for for that. The callback function, along with the structure
> containing the response data, provide information whether more data is expected
> to receive or not. A very simple response handling function is presented below:

Reviewer 2: ```suggestion
If the server responds to the request, the response will be provided to the
application through the response callback registered in the request structure.
The response may be provided in chunks, therefore the application shall be
prepared for that. The callback function, along with the structure
containing the response data, provides information about whether more data is expected
to be received. A very simple response handling function is presented below:
```

> If the server responds to the request, the response will be provided to the
> application through the response callback registered in the request structure.
> The response may be provided in chunks, therefore the application shall be
> prepared for that. The callback function, along with the structure containing
> the response data, provides information about whether more data is expected
> to be received. A very simple response handling function is presented below:

Reviewer: ```suggestion
If the server responds to the request, the library provides the response to the
application through the response callback registered in the request structure.
As the library can provide the response in chunks, the application must be able
to process these.

Together with the structure containing the response data, the callback function
also provides information about whether the library expects to receive more data.

The following is an example of a very simple response handling function:
```
